File: src/copilot_export/parser.py
"""Parse a Copilot CLI session directory into a ParsedSession."""

from __future__ import annotations

import json
import re
import sqlite3
from pathlib import Path
from typing import Any

import yaml

from .models import (
    Checkpoint,
    ConversationTurn,
    Event,
    EventType,
    ParsedSession,
    SubAgentRun,
    Todo,
    TodoDep,
    ToolCall,
    ToolRequest,
    ToolResult,
    WorkspaceMetadata,
)

# Regex to detect thinking blocks in assistant content
_THINKING_RE = re.compile(
    r"<(?:antml:)?thinking(?:_mode)?[^>]*>(.*?)</(?:antml:)?thinking(?:_mode)?>",
    re.DOTALL,
)


def parse_session(session_dir: str | Path) -> ParsedSession:
    """Parse all files in a session directory and return a structured ParsedSession."""
    root = Path(session_dir)
    if not root.is_dir():
        raise FileNotFoundError(f"Session directory not found: {root}")

    workspace = _parse_workspace(root / "workspace.yaml")
    events = _parse_events(root / "events.jsonl")
    todos, todo_deps = _parse_db(root / "session.db")
    plan = _read_text(root / "plan.md")
    checkpoints = _parse_checkpoints(root / "checkpoints")

    copilot_version = ""
    shutdown_stats: dict[str, Any] = {}
    errors: list[dict[str, Any]] = []

    for ev in events:
        if ev.type == EventType.SESSION_START:
            copilot_version = ev.data.get("copilotVersion", "")
        elif ev.type == EventType.SESSION_SHUTDOWN:
            shutdown_stats = ev.data
        elif ev.type == EventType.SESSION_ERROR:
            errors.append(ev.data)

    turns = _build_turns(events)

    return ParsedSession(
        workspace=workspace,
        events=events,
        turns=turns,
        todos=todos,
        todo_deps=todo_deps,
        checkpoints=checkpoints,
        plan=plan,
        copilot_version=copilot_version,
        shutdown_stats=shutdown_stats,
        errors=errors,
        session_dir=root,
    )


# ---------------------------------------------------------------------------
# File parsers
# ---------------------------------------------------------------------------

def _parse_workspace(path: Path) -> WorkspaceMetadata:
    if not path.exists():
        return WorkspaceMetadata(id="", cwd="")
    with open(path, "r", encoding="utf-8") as f:
        data = yaml.safe_load(f) or {}
    return WorkspaceMetadata.from_dict(data)


def _parse_events(path: Path) -> list[Event]:
    if not path.exists():
        return []
    events: list[Event] = []
    with open(path, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            try:
                obj = json.loads(line)
                events.append(Event.from_dict(obj))
            except json.JSONDecodeError:
                continue
    return events


def _parse_db(path: Path) -> tuple[list[Todo], list[TodoDep]]:
    if not path.exists():
        return [], []
    todos: list[Todo] = []
    deps: list[TodoDep] = []
    try:
        conn = sqlite3.connect(str(path))
        cursor = conn.cursor()

        # Check if tables exist
        cursor.execute(
            "SELECT name FROM sqlite_master WHERE type='table' AND name='todos'"
        )
        if cursor.fetchone():
            cursor.execute("SELECT id, title, description, status FROM todos")
            for row in cursor.fetchall():
                todos.append(Todo(id=row[0], title=row[1], description=row[2] or "", status=row[3] or "pending"))

        cursor.execute(
            "SELECT name FROM sqlite_master WHERE type='table' AND name='todo_deps'"
        )
        if cursor.fetchone():
            cursor.execute("SELECT todo_id, depends_on FROM todo_deps")
            for row in cursor.fetchall():
                deps.append(TodoDep(todo_id=row[0], depends_on=row[1]))

        conn.close()
    except sqlite3.Error:
        pass
    return todos, deps


def _parse_checkpoints(cp_dir: Path) -> list[Checkpoint]:
    if not cp_dir.is_dir():
        return []
    index_path = cp_dir / "index.md"
    checkpoints: list[Checkpoint] = []
    if index_path.exists():
        # Parse the index table to get checkpoint order and titles
        content = index_path.read_text(encoding="utf-8")
        # Matches table rows like: | 1 | Title | filename.md |
        for match in re.finditer(
            r"\|\s*(\d+)\s*\|\s*(.+?)\s*\|\s*(.+?)\s*\|", content
        ):
            idx = int(match.group(1))
            title = match.group(2).strip()
            filename = match.group(3).strip()
            cp_path = cp_dir / filename
            cp_content = ""
            if cp_path.exists():
                cp_content = cp_path.read_text(encoding="utf-8")
            checkpoints.append(
                Checkpoint(index=idx, title=title, filename=filename, content=cp_content)
            )
    else:
        # Fallback: read all .md files in order
        for i, md_file in enumerate(sorted(cp_dir.glob("*.md")), 1):
            if md_file.name == "index.md":
                continue
            checkpoints.append(
                Checkpoint(
                    index=i,
                    title=md_file.stem,
                    filename=md_file.name,
                    content=md_file.read_text(encoding="utf-8"),
                )
            )
    return checkpoints


def _read_text(path: Path) -> str:
    if not path.exists():
        return ""
    return path.read_text(encoding="utf-8")


# ---------------------------------------------------------------------------
# Build conversation turns from events
# ---------------------------------------------------------------------------

def _build_turns(events: list[Event]) -> list[ConversationTurn]:
    """Group events into a chronological list of conversation turns."""

    # Index tool execution events by toolCallId for quick lookup
    tool_starts: dict[str, Event] = {}
    tool_completes: dict[str, Event] = {}
    subagent_starts: dict[str, Event] = {}
    subagent_ends: dict[str, Event] = {}  # completed or failed

    for ev in events:
        if ev.type == EventType.TOOL_EXECUTION_START:
            tool_starts[ev.data.get("toolCallId", "")] = ev
        elif ev.type == EventType.TOOL_EXECUTION_COMPLETE:
            tool_completes[ev.data.get("toolCallId", "")] = ev
        elif ev.type == EventType.SUBAGENT_STARTED:
            subagent_starts[ev.data.get("toolCallId", "")] = ev
        elif ev.type in (EventType.SUBAGENT_COMPLETED, EventType.SUBAGENT_FAILED):
            subagent_ends[ev.data.get("toolCallId", "")] = ev

    turns: list[ConversationTurn] = []
    current_mode: str | None = None
    current_assistant_turn: ConversationTurn | None = None

    for ev in events:
        if ev.type == EventType.SESSION_MODE_CHANGED:
            current_mode = ev.data.get("newMode", current_mode)

        elif ev.type == EventType.USER_MESSAGE:
            turns.append(
                ConversationTurn(
                    role="user",
                    timestamp=ev.timestamp,
                    content=ev.data.get("content", ""),
                    mode=current_mode,
                )
            )

        elif ev.type == EventType.ASSISTANT_TURN_START:
            current_assistant_turn = ConversationTurn(
                role="assistant",
                timestamp=ev.timestamp,
                mode=current_mode,
            )

        elif ev.type == EventType.ASSISTANT_MESSAGE:
            if current_assistant_turn is None:
                current_assistant_turn = ConversationTurn(
                    role="assistant", timestamp=ev.timestamp, mode=current_mode
                )

            raw_content = ev.data.get("content", "")

            # Extract thinking sections
            thinking_parts = _THINKING_RE.findall(raw_content)
            if thinking_parts:
                current_assistant_turn.thinking += "\n".join(thinking_parts)
                # Remove thinking from displayed content
                clean_content = _THINKING_RE.sub("", raw_content).strip()
            else:
                clean_content = raw_content

            if clean_content:
                if current_assistant_turn.content:
                    current_assistant_turn.content += "\n\n" + clean_content
                else:
                    current_assistant_turn.content = clean_content

            # Process tool requests
            for tr_data in ev.data.get("toolRequests", []):
                tr = ToolRequest.from_dict(tr_data)

                # Skip internal-only tools
                if tr.name in ("report_intent",):
                    continue

                tc_id = tr.tool_call_id
                result = None
                if tc_id in tool_starts and tc_id in tool_completes:
                    result = ToolResult.from_execution_events(
                        tool_starts[tc_id], tool_completes[tc_id]
                    )

                description = tr.arguments.get("description", "")

                # Check if this is a sub-agent call
                if tc_id in subagent_starts:
                    sa_start = subagent_starts[tc_id]
                    sa_end = subagent_ends.get(tc_id)
                    sa = SubAgentRun(
                        tool_call_id=tc_id,
                        agent_name=sa_start.data.get("agentName", ""),
                        display_name=sa_start.data.get("agentDisplayName", ""),
                        description=sa_start.data.get("agentDescription", ""),
                        success=sa_end.type == EventType.SUBAGENT_COMPLETED if sa_end else True,
                        error=sa_end.data.get("error", "") if sa_end and sa_end.type == EventType.SUBAGENT_FAILED else "",
                    )
                    current_assistant_turn.sub_agents.append(sa)
                else:
                    current_assistant_turn.tool_calls.append(
                        ToolCall(request=tr, result=result, description=description)
                    )

        elif ev.type == EventType.ASSISTANT_TURN_END:
            if current_assistant_turn is not None:
                turns.append(current_assistant_turn)
                current_assistant_turn = None

    # Flush any unfinished assistant turn
    if current_assistant_turn is not None:
        turns.append(current_assistant_turn)

    return turns

File: go-cli/parser.go
package main

import (
	"bufio"
	"database/sql"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"

	"gopkg.in/yaml.v3"
	_ "modernc.org/sqlite"
)

// Regex to detect thinking blocks in assistant content.
var thinkingRE = regexp.MustCompile(`(?s)<(?:antml:)?thinking(?:_mode)?[^>]*>(.*?)</(?:antml:)?thinking(?:_mode)?>`)

// ParseSession parses all files in a session directory.
func ParseSession(sessionDir string) (*ParsedSession, error) {
	info, err := os.Stat(sessionDir)
	if err != nil || !info.IsDir() {
		return nil, fmt.Errorf("session directory not found: %s", sessionDir)
	}

	workspace := parseWorkspace(filepath.Join(sessionDir, "workspace.yaml"))
	events := parseEvents(filepath.Join(sessionDir, "events.jsonl"))
	todos, todoDeps := parseDB(filepath.Join(sessionDir, "session.db"))
	plan := readText(filepath.Join(sessionDir, "plan.md"))
	checkpoints := parseCheckpoints(filepath.Join(sessionDir, "checkpoints"))

	copilotVersion := ""
	var shutdownStats map[string]interface{}
	var errors []map[string]interface{}

	for _, ev := range events {
		switch ev.Type {
		case EventSessionStart:
			if v, ok := ev.Data["copilotVersion"].(string); ok {
				copilotVersion = v
			}
		case EventSessionShutdown:
			shutdownStats = ev.Data
		case EventSessionError:
			errors = append(errors, ev.Data)
		}
	}

	turns := buildTurns(events)

	return &ParsedSession{
		Workspace:      workspace,
		Events:         events,
		Turns:          turns,
		Todos:          todos,
		TodoDeps:       todoDeps,
		Checkpoints:    checkpoints,
		Plan:           plan,
		CopilotVersion: copilotVersion,
		ShutdownStats:  shutdownStats,
		Errors:         errors,
		SessionDir:     sessionDir,
	}, nil
}

func parseWorkspace(path string) WorkspaceMetadata {
	data, err := os.ReadFile(path)
	if err != nil {
		return WorkspaceMetadata{}
	}
	var ws WorkspaceMetadata
	if err := yaml.Unmarshal(data, &ws); err != nil {
		return WorkspaceMetadata{}
	}
	return ws
}

func parseEvents(path string) []Event {
	f, err := os.Open(path)
	if err != nil {
		return nil
	}
	defer f.Close()

	var events []Event
	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 0, 1024*1024), 10*1024*1024)

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var raw map[string]interface{}
		if err := json.Unmarshal([]byte(line), &raw); err != nil {
			continue
		}

		ev := Event{
			Type:     strVal(raw, "type"),
			ID:       strVal(raw, "id"),
			RawTS:    strVal(raw, "timestamp"),
			ParentID: strVal(raw, "parentId"),
		}
		if d, ok := raw["data"].(map[string]interface{}); ok {
			ev.Data = d
		} else {
			ev.Data = make(map[string]interface{})
		}
		ev.Timestamp = ParseTS(ev.RawTS)
		events = append(events, ev)
	}
	return events
}

func parseDB(path string) ([]Todo, []TodoDep) {
	if _, err := os.Stat(path); err != nil {
		return nil, nil
	}

	db, err := sql.Open("sqlite", path)
	if err != nil {
		return nil, nil
	}
	defer db.Close()

	var todos []Todo
	var deps []TodoDep

	if tableExists(db, "todos") {
		rows, err := db.Query("SELECT id, title, description, status FROM todos")
		if err == nil {
			defer rows.Close()
			for rows.Next() {
				var t Todo
				var desc, status sql.NullString
				if err := rows.Scan(&t.ID, &t.Title, &desc, &status); err == nil {
					t.Description = desc.String
					t.Status = status.String
					if t.Status == "" {
						t.Status = "pending"
					}
					todos = append(todos, t)
				}
			}
		}
	}

	if tableExists(db, "todo_deps") {
		rows, err := db.Query("SELECT todo_id, depends_on FROM todo_deps")
		if err == nil {
			defer rows.Close()
			for rows.Next() {
				var d TodoDep
				if err := rows.Scan(&d.TodoID, &d.DependsOn); err == nil {
					deps = append(deps, d)
				}
			}
		}
	}

	return todos, deps
}

func tableExists(db *sql.DB, name string) bool {
	var n string
	err := db.QueryRow("SELECT name FROM sqlite_master WHERE type='table' AND name=?", name).Scan(&n)
	return err == nil
}

func parseCheckpoints(cpDir string) []Checkpoint {
	info, err := os.Stat(cpDir)
	if err != nil || !info.IsDir() {
		return nil
	}

	indexPath := filepath.Join(cpDir, "index.md")
	var checkpoints []Checkpoint

	if content, err := os.ReadFile(indexPath); err == nil {
		// Parse table rows: | idx | title | filename |
		re := regexp.MustCompile(`\|\s*(\d+)\s*\|\s*(.+?)\s*\|\s*(.+?)\s*\|`)
		matches := re.FindAllStringSubmatch(string(content), -1)
		for _, m := range matches {
			idx, _ := strconv.Atoi(m[1])
			title := strings.TrimSpace(m[2])
			filename := strings.TrimSpace(m[3])
			cpContent := readText(filepath.Join(cpDir, filename))
			checkpoints = append(checkpoints, Checkpoint{
				Index:    idx,
				Title:    title,
				Filename: filename,
				Content:  cpContent,
			})
		}
	} else {
		// Fallback: read all .md files
		entries, _ := os.ReadDir(cpDir)
		var mdFiles []string
		for _, e := range entries {
			if !e.IsDir() && strings.HasSuffix(e.Name(), ".md") && e.Name() != "index.md" {
				mdFiles = append(mdFiles, e.Name())
			}
		}
		sort.Strings(mdFiles)
		for i, name := range mdFiles {
			stem := strings.TrimSuffix(name, ".md")
			checkpoints = append(checkpoints, Checkpoint{
				Index:    i + 1,
				Title:    stem,
				Filename: name,
				Content:  readText(filepath.Join(cpDir, name)),
			})
		}
	}
	return checkpoints
}

func readText(path string) string {
	data, err := os.ReadFile(path)
	if err != nil {
		return ""
	}
	return string(data)
}

// buildTurns groups events into conversation turns.
func buildTurns(events []Event) []ConversationTurn {
	// Index tool/subagent events by toolCallId
	toolStarts := map[string]Event{}
	toolCompletes := map[string]Event{}
	subagentStarts := map[string]Event{}
	subagentEnds := map[string]Event{}

	for _, ev := range events {
		tcID := strVal(ev.Data, "toolCallId")
		switch ev.Type {
		case EventToolExecutionStart:
			toolStarts[tcID] = ev
		case EventToolExecutionComplete:
			toolCompletes[tcID] = ev
		case EventSubagentStarted:
			subagentStarts[tcID] = ev
		case EventSubagentCompleted, EventSubagentFailed:
			subagentEnds[tcID] = ev
		}
	}

	var turns []ConversationTurn
	var currentMode string
	var cur *ConversationTurn

	for _, ev := range events {
		switch ev.Type {
		case EventSessionModeChanged:
			if m, ok := ev.Data["newMode"].(string); ok {
				currentMode = m
			}

		case EventUserMessage:
			content, _ := ev.Data["content"].(string)
			turns = append(turns, ConversationTurn{
				Role:      "user",
				Timestamp: ev.Timestamp,
				Content:   content,
				Mode:      currentMode,
			})

		case EventAssistantTurnStart:
			cur = &ConversationTurn{
				Role:      "assistant",
				Timestamp: ev.Timestamp,
				Mode:      currentMode,
			}

		case EventAssistantMessage:
			if cur == nil {
				cur = &ConversationTurn{
					Role:      "assistant",
					Timestamp: ev.Timestamp,
					Mode:      currentMode,
				}
			}

			rawContent, _ := ev.Data["content"].(string)

			// Extract thinking sections
			thinkingParts := thinkingRE.FindAllStringSubmatch(rawContent, -1)
			if len(thinkingParts) > 0 {
				for _, m := range thinkingParts {
					if cur.Thinking != "" {
						cur.Thinking += "\n"
					}
					cur.Thinking += m[1]
				}
				cleanContent := strings.TrimSpace(thinkingRE.ReplaceAllString(rawContent, ""))
				if cleanContent != "" {
					if cur.Content != "" {
						cur.Content += "\n\n" + cleanContent
					} else {
						cur.Content = cleanContent
					}
				}
			} else if rawContent != "" {
				if cur.Content != "" {
					cur.Content += "\n\n" + rawContent
				} else {
					cur.Content = rawContent
				}
			}

			// Process tool requests
			if toolReqs, ok := ev.Data["toolRequests"].([]interface{}); ok {
				for _, trRaw := range toolReqs {
					trMap, ok := trRaw.(map[string]interface{})
					if !ok {
						continue
					}
					tr := ToolRequest{
						ToolCallID: strVal(trMap, "toolCallId"),
						Name:       strVal(trMap, "name"),
					}
					if args, ok := trMap["arguments"].(map[string]interface{}); ok {
						tr.Arguments = args
					} else {
						tr.Arguments = make(map[string]interface{})
					}

					// Skip internal-only tools
					if tr.Name == "report_intent" {
						continue
					}

					tcID := tr.ToolCallID
					var result *ToolResult
					if _, hasStart := toolStarts[tcID]; hasStart {
						if _, hasComplete := toolCompletes[tcID]; hasComplete {
							r := ToolResultFromEvents(toolStarts[tcID], toolCompletes[tcID])
							result = &r
						}
					}

					desc, _ := tr.Arguments["description"].(string)

					// Check if this is a sub-agent call
					if saStart, isSA := subagentStarts[tcID]; isSA {
						sa := SubAgentRun{
							ToolCallID:  tcID,
							AgentName:   strVal(saStart.Data, "agentName"),
							DisplayName: strVal(saStart.Data, "agentDisplayName"),
							Description: strVal(saStart.Data, "agentDescription"),
							Success:     true,
						}
						if saEnd, hasEnd := subagentEnds[tcID]; hasEnd {
							sa.Success = saEnd.Type == EventSubagentCompleted
							if saEnd.Type == EventSubagentFailed {
								sa.Error, _ = saEnd.Data["error"].(string)
							}
						}
						cur.SubAgents = append(cur.SubAgents, sa)
					} else {
						cur.ToolCalls = append(cur.ToolCalls, ToolCall{
							Request:     tr,
							Result:      result,
							Description: desc,
						})
					}
				}
			}

		case EventAssistantTurnEnd:
			if cur != nil {
				turns = append(turns, *cur)
				cur = nil
			}
		}
	}

	// Flush any unfinished assistant turn
	if cur != nil {
		turns = append(turns, *cur)
	}

	return turns
}

// strVal safely extracts a string from a map.
func strVal(m map[string]interface{}, key string) string {
	if v, ok := m[key].(string); ok {
		return v
	}
	return ""
}
